host/docs/rfnoc_modtool.1
.TH "rfnoc_modtool" 1 "4.8.0" UHD "User Commands"
.SH NAME
rfnoc_modtool - RFNoC OOT module management tool

.SH DESCRIPTION
Create and manage RFNoC OOT modules.


.SH SYNOPSIS
.B  rfnoc_modtool [COMMAND] [OPTIONS]

.SH COMMANDS
Run rfnoc_modtool COMMAND --help for more information on a specific command.

.IP "create"
Create a new RFNoC OOT module.
.IP "add"
Add a new block to an existing RFNoC OOT module.
.IP "add-groot"
Add GNU Radio OOT Module.
.IP "add-grc"
Add GNU Radio GRC bindings for an existing block in this RFNoC OOT module.
.IP "make-yaml"
YAML Creation Wizard -- Generate an RFNoC block YAML descriptor file based on user input.

.SH CREATING NEW MODULES
.sp
When running rfnoc_modtool create, a new directory will be created that contains
the necessary files for a new RFNoC OOT module. This directory will only contain
boilerplate code, and the user will need to add their own blocks to the module.

By calling 'rfnoc_modtool add-groot', the user can add a GNU Radio OOT module
inside the RFNoC OOT module.

.SH ADDING BLOCKS
.sp
When running rfnoc_modtool add, a new block will be added to an existing RFNoC OOT module.
This requires previously having run rfnoc_modtool create to create the module.

Note that in order for 'rfnoc_modtool add' to succeed, the user must provide a
block descriptor file in YAML format. This file describes the block and its parameters.
The user can create this file manually, or use the YAML Creation Wizard by running
rfnoc_modtool make-yaml. The wizard will prompt the user for the necessary information
to create a valid YAML file.

Once blocks have been created, the user can add GNU Radio GRC bindings for the blocks
by running 'rfnoc_modtool add-grc'.


.SH EXAMPLES

.SS Create a new RFNoC OOT module called "filter"
.sp
rfnoc_modtool create filter
.ft

.SS Add a block called "fir" to the "filter" module (this requires having create a fir.yml block descriptor file)
.sp
rfnoc_modtool -C ./rfnoc-filter add fir -y rfnoc-foo/rfnoc/blocks/fir.yml
.ft

.fi

.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
rfnoc_image_builder(1)
.SH AUTHOR
This manual page was written by Martin Braun.
.SH COPYRIGHT
Copyright (c) 2015-2022 Ettus Research, A National Instruments Brand
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 004.next

* New Features
  - X3x0: Add support for the gps_iface feature. This allows sending arbitrary
    commands to the GPS module.
  - rfnoc_modtool: Add a new subcommand, `make-yaml`, that generates a block
    YAML file from some user inputs. When calling `rfnoc_modtool add`  without
    a valid block YAML file, the tool will now automatically call `make-yaml` to
    generate a new YAML file.
* API Changes
  - `uhd::math::gcd` and `uhd::math::lcm` are removed from the public
     headers. Use `std::gcd` and `std::lcm` instead if C++17 is available,
     or `boost::math::gcd` and `boost::math::lcm` if not.
  - When calling `recv()` on an RX streamer connected to an RFNoC device, we
    no longer artificially inflate the timeout value. The old behaviour was
    in some cases abused to detect overruns, but this is not a reliable method
    and is now explicitly declared unsupported.
* Bug Fixes
  - RFNoC
    - Calling `recv()` on an RX streamer connected to an RFNoC device with a
      zero timeout value now correctly returns immediately.
    - rfnoc-gain: Moved the fpga/ subdirectory under rfnoc/. This way, the
      directory structure inside the OOT module is identical to the directory
      structure after installation.
  - X410:
    - Fixed NCO reset mechanism that caused ADC self-cal errors to occur during
      reopening of a UHD session when timed tune commands were previously used.
* Removals / Deprecations
  - Removed tools/uhd_dump. This tool is obsolete, the USRP X3x0 is not even
    compatible with it any more. Dumps can be recorded and analyzed with wireshark
    and the UHD dissectors in here.


## 004.008.000.000
* Highlights / Main Changes
  - Addition of rfnoc_modtool and further improvements to RFNoC Image Builder
    toolsets. This further improves the ability to design RFNoC blocks and
    create custom FPGA bitfiles.
  - Replaced RFNoC FFT block with a new version
  - Improved Windows installation capabilities including distribution
    of binary Python bindings for Windows though PyPI
* New Features
  - Image Builder
    - Add GRC support. This allows designing RFNoC bitfiles from GNU Radio
      Companion (GRC).
    - Improve clock connection checks and checks for duplicate connections.
    - Improve IO port compat check (e.g., check if wire widths match).
    - Allow default clocks in domain checks.
    - Simplify image core YAMLs by better usage of RADIO_NIPC parameter.
    - Improve parameter resolution.
    - Improve error messages for missing YAML files.
    - Add --SYNTH and --CHECK options.
    - Add support for building an FPGA image using multiple parallel jobs and
      unique seeds with repeat_fpga_build.py and the --fpga-jobs option.
    - Add 'q' (or 'quote') filter to YAML files to allow quoting strings.
  - rfnoc-gain (OOT RFNoC example)
    - Overhaul directory structure, and rename to rfnoc-gain from
      rfnoc-example.
    - Simplify dynamic loading of OOT DLLs by using the new modules.d feature.
    - Add CE clock support to gain block. This enables the example on X4x0.
    - Add a GNU Radio subdirectory with examples of how to run the gain block
      in GNU Radio.
  - Add rfnoc_modtool. This is a command line utility to help design OOT RFNoC
    blocks and replaces the previously available blocktool.
  - RFNoC
    - Add tune requests. This allows tuning a complete graph as known from
      multi_usrp instead of single blocks individually.
    - Replace FFT block with a new version supporting larger FFT sizes and
      cyclic prefix insertion and removal.
  - multi_usrp
    - Add Python bindings for get_user_settings_iface()
  - General UHD
    - Add modules.d support
    - Add uhd::find_uhd_command() API call
    - Remove old-style Boost.Asio usage for compatibility with future versions
      of Boost.
  - Embedded Filesystem
    - Support for USB Audio devices on E31X, E320, N3XX and X4XX
  - Windows binary installer
    - Simplified the Windows installer experience by reducing the number of
      individual downloads from 4 (for USB devices) to 2, by including
      libusb.dll and USB driver installer files into NSIS-built installer
      binaries.
    - Add new cmake feature option called ENABLE_EXTEND_WIN_PACKAGING,
      which defaults to off. Enabling this feature will cause a number
      of convenience components to be included into the Windows deployment
      package (aka. installer).
      Enabling this feature has no effect on other platforms.
    - Added binary Python bindings for Windows installable through PyPI
      (pip install uhd)
  - Documentation
    - Add documentation about relationship of MTU, samples per packet and
      atomic item size
* Dependencies changes
  - Increase minimum required versions
    - CMake to 3.12
    - Boost to 1.66
* Bug Fixes
  - General UHD
    - Fix compatibility with DPDK >= 22.11
    - Fix compiler warnings for better compatibility with C++17 and 20.
    - Add logic for loading uhd.dll from the correct path for Windows with
      Python3.8+.
    - Add logic to warn about potential incompatibilities when importing
      Python bindings.
    - Remove duplicate results from find
    - Release GIL when calling find from Python which improves response time
      for large setups.
  - Python support
    - On Windows now also enable building python support if minimum supported
      python version is detected.
    - Changed version number format for build Windows binaries to common
      major.api.abi.patch format.
  - MPM
    - Allow images without RF frontend (will only be initialized if FPGA
      reports availability)
    - Fix usrp_update_fs to work for all embedded devices (and not just X4xx)
  - E320: Ensure consistent sequencing when powering on/off GPSDO
  - N310/N300: Fix LO query API for lowband LO
  - RFNoC
    - RFNoC DDC/DUC block (used in all Gen-3 USRPs and X410): Fix fractional
      frequency offset.

host/python/uhd/rfnoc_utils/rfnoc_modtool.py
#!/usr/bin/env python3
"""RFNoC Modtool: The tool to create and manipulate RFNoC OOT modules."""
#
# Copyright 2024 Ettus Research, a National Instruments Brand
#
# SPDX-License-Identifier: GPL-3.0-or-later
#

import argparse
import glob
import os
import sys

from ruamel.yaml import YAML
from uhd import get_pkg_data_path

from .log import init_logging
from .step_executor import StepExecutor
from .utils import resolve


def get_command_repo_dir():
    """Return the path to the command repository directory (where the command YAMLS are stored)."""
    this_dir = os.path.dirname(os.path.realpath(__file__))
    return os.path.join(this_dir, "modtool_commands")


def collect_commands():
    """Load all available commands.

    The return value is a dictionary where the keys are the command names and
    the values are raw contents of the YAML as Python structures.
    """
    yamls = glob.glob(os.path.join(get_command_repo_dir(), "*.yml"))
    yaml = YAML()
    commands = {}
    for y in yamls:
        cmd_name = os.path.basename(y).replace(".yml", "")
        with open(y, "r", encoding="utf-8") as f:
            commands[cmd_name] = yaml.load(f)
    return commands


def parse_args(cmds):
    """Parse args.

    This will dynamically populate main- and subparsers from the YAML configs
    given in cmds.
    """
    arg_parser = argparse.ArgumentParser(description=__doc__)
    arg_parser.add_argument(
        "-C", "--directory", help="Change to this directory before running the command"
    )
    arg_parser.add_argument(
        "-l", "--log-level", default="INFO", help="Set the log level (default: INFO)"
    )
    subparsers = arg_parser.add_subparsers(dest="command", required=True)
    parser_list = []
    for cmd in cmds:
        new_parser = subparsers.add_parser(cmd, help=cmds[cmd]["help"])
        for arg_name, arg_info in cmds[cmd]["args"].items():
            name_or_flags = (
                f"{arg_name}" if "name_or_flags" not in arg_info else arg_info.pop("name_or_flags")
            )
            if not isinstance(name_or_flags, list):
                name_or_flags = [name_or_flags]
            # We don't want to use eval(), but for most types we need to do
            # just that, so manually map
            if "type" in arg_info:
                try:
                    arg_info["type"] = {
                        "str": str,
                        "int": int,
                        "float": float,
                    }[arg_info["type"]]
                except KeyError:
                    pass
            new_parser.add_argument(*name_or_flags, **arg_info)
        parser_list.append(new_parser)
    return arg_parser.parse_args()


def get_global_vars(pkg_data_dir):
    """Create a dictionary with global variables for the key resolution."""
    return {
        "RFNOC_PKG_DIR": pkg_data_dir,
        "CWD": os.getcwd(),
    }


def resolve_vars(cmd, global_vars, args):
    """Resolve all variables in the command."""
    for var_name, var_val in cmd.get("variables", {}).items():
        cmd["variables"][var_name] = resolve(
            var_val, args=args, **global_vars, **cmd.get("variables", {})
        )
    return cmd


def check_valid_oot_dir(oot_dir):
    """Check if the given directory contains a valid OOT module."""
    return os.path.isfile(os.path.join(oot_dir, "CMakeLists.txt")) and os.path.isdir(
        os.path.join(oot_dir, "rfnoc")
    )


def main():
    """Run main rfnoc_modtool function."""
    pkg_data_dir = os.path.normpath(get_pkg_data_path())
    cmds = collect_commands()
    args = parse_args(cmds)
    init_logging(log_level=args.log_level)
    if args.directory:
        os.chdir(args.directory)
    cmd = cmds[args.command]
    global_vars = get_global_vars(pkg_data_dir)
    if not cmd.get("skip_identify_module", False):
        if not check_valid_oot_dir(os.getcwd()):
            print("Error: Not a valid OOT module directory")
            return 1
        oot_dir_name = os.path.split(os.getcwd())[-1]
        module_name = oot_dir_name.replace("rfnoc-", "")
        global_vars["MODULE_NAME"] = module_name
        global_vars["MODULE_NAME_FULL"] = oot_dir_name
    cmd = resolve_vars(cmd, global_vars, args)
    executor = StepExecutor(global_vars, args, cmd)
    executor.run(cmd["steps"])
    return 0


if __name__ == "__main__":
    sys.exit(main())
